File: Virtual Machine/Nova.VirtualMachine/Nova.VirtualMachine/ByteField.cpp
#include "ByteField.h"
#include "Exec.h"
#include "Logger.h"

void ByteField::Deserialize(BinaryReader& reader)
{
	this->name = reader.ReadString();
	this->modifiers = (Modifiers)reader.Read<char>();
	this->valueBlock = new ByteBlock();
	this->valueBlock->Deserialize(reader);
}

void ByteField::Dispose()
{
	delete valueBlock;
}

/*
	RuntimeContext.GetExecutingClass() will return null here. It is a problem?
*/
void ByteField::Initializer(RuntimeContext* context)
{
	/*std::vector<RuntimeContext::RuntimeElement> locales;
	Exec::Execute(context, locales, this->valueBlock->instructions);

	if (context->GetStackSize() == 1)
	{
		this->value = context->PopStack();
	}
	else
	{
		Logger::Error("Unable to initialize field " + this->name + " invalid value.");
	} */
}

File: Compiler/Nova.VirtualMachine/Members/ByteField.cs
using Nova.Utils;
using Nova.Utils.IO;
using Nova.VirtualMachine.Runtime;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Nova.VirtualMachine.Members
{
    public class ByteField
    {
        public string Name
        {
            get;
            private set;
        }
        public Modifiers Modifiers
        {
            get;
            private set;
        }
        public ByteBlock ValueBlock
        {
            get;
            set;
        }
        public object Value
        {
            get;
            set;
        }
        public ByteField()
        {
            ValueBlock = new ByteBlock();
        }
        public void Deserialize(CppBinaryReader reader)
        {
            this.Name = reader.ReadString();
            this.Modifiers = (Modifiers)reader.ReadByte();

            ValueBlock.Deserialize(reader);
        }

        public void Initializer(RuntimeContext context)
        {
            Exec.Execute(context, null, this.ValueBlock.Instructions);

            if (context.GetStackSize() == 1)
            {
                this.Value = context.PopStack();
            }
            else
            {
                Logger.Write("Unable to initialize field " + Name + " invalid value.", LogType.Error);
            }
        }
    }
}
